File: 1718/a.py
import sys
import string
from util import ints


def main():
    def get(value_or_register):
        if value_or_register in registers:
            return registers[value_or_register]
        return value_or_register

    f = open(sys.argv[1] if len(sys.argv) > 1 else 'in')
    lines = [l.rstrip('\n') for l in f]
    program = [ints(line.split()) for line in lines]

    registers = { r: 0 for r in string.ascii_lowercase }
    ip = 0

    while 0 <= ip < len(program):
        instruction = program[ip]
        op, *args = instruction
        if op == 'snd':
            x = args[0]
            last_sound = get(x)
        elif op == 'set':
            x, y = args
            registers[x] = get(y)
        elif op == 'add':
            x, y = args
            registers[x] += get(y)
        elif op == 'mul':
            x, y = args
            registers[x] *= get(y)
        elif op == 'mod':
            x, y = args
            registers[x] %= get(y)
        elif op == 'rcv':
            x = args[0]
            if get(x):
                print(last_sound)
                exit()
        elif op == 'jgz':
            x, y = args
            if get(x):
                ip += get(y)
                continue
        else:
            1/0
        ip += 1


if __name__ == '__main__':
    main()

File: 1718/a.js
const fs = require('fs');

function main() {
  function get(valueOrRegister) {
    const x = valueOrRegister;
    if (+x === x) { // i.e. is a number
      return x;
    } else {
      return registers[x];
    }
  }

  // Parse input file
  const program =
    fs.readFileSync('in', 'utf-8')
      .trim()
      .split('\n')
      .map(line => // e.g. "set i 31"
        line
          .split(' ')
          .map(word => // e.g. "set" or "i" or "31" (string)
            Number.isNaN(+word)
            ? word
            : +word
          )
      )
      .map(([command, x, y]) => // e.g. ["set", "i", 31]
        ({command, x, y}) // e.g. {command: "set", x: "i", y: 31}
      );

  // Run program
  let ip = 0;
  const registers = Object.fromEntries(
    Array.from('abcdefghijklmnopqrstuvwxyz')
      .map(letter => [letter, 0])
  );
  let lastSound;
  while (0 <= ip && ip < program.length) {
    const { command, x, y } = program[ip];
    if (command === 'snd') {
      lastSound = get(x);
    } else if (command === 'set') {
      registers[x] = get(y);
    } else if (command === 'add') {
      registers[x] += get(y);
    } else if (command === 'mul') {
      registers[x] *= get(y);
    } else if (command === 'mod') {
      registers[x] %= get(y);
    } else if (command === 'rcv') {
      if (get(x) !== 0) {
        console.log(lastSound);
        break;
      }
    } else if (command === 'jgz') {
      if (get(x) > 0) {
        ip += get(y);
        continue;
      }
    } else {
      throw new Error('Invalid command');
    }
    ip += 1;
  }
}

main();
